server/src/test/java/org/elasticsearch/cluster/coordination/AtomicRegisterCoordinatorTests.java

```diff
@@ -0,0 +1,614 @@
+        // into the register and therefore the no master block won't be applied
+    }
+
+    @Override
+    @AwaitsFix(bugUrl = "ES-5645")
+    public void testAppliesNoMasterBlockWritesIfConfigured() {
+        // If the disconnected node is the leader it will continue to have connectivity
+        // into the register and therefore the no master block won't be applied
+    }
+
+    @Override
+    @AwaitsFix(bugUrl = "ES-5645")
+    public void testAppliesNoMasterBlockAllIfConfigured() {
+        // If the disconnected node is the leader it will continue to have connectivity
+        // into the register and therefore the no master block won't be applied
+    }
+
+    @Override
+    @AwaitsFix(bugUrl = "ES-5645")
+    public void testAppliesNoMasterBlockMetadataWritesIfConfigured() {
+        // If the disconnected node is the leader it will continue to have connectivity
+        // into the register and therefore the no master block won't be applied
+    }
+
+    @Override
+    @AwaitsFix(bugUrl = "ES-5645")
+    public void testClusterCannotFormWithFailingJoinValidation() {
+        // A single node can form a cluster in this case
+    }
+
+    @Override
+    @AwaitsFix(bugUrl = "ES-5645")
+    public void testCannotJoinClusterWithDifferentUUID() {
+        // The cluster2 leader is considered dead since we only run the nodes in cluster 1
+        // therefore the node coming from cluster 2 ends up taking over the old master in cluster 2
+        // TODO: add more checks to avoid forming a mixed cluster between register based and traditional clusters
+    }
+
+    @Override
+    public void testJoiningNodeReceivesFullState() {
+        try (Cluster cluster = new Cluster(randomIntBetween(1, 5))) {
+            cluster.runRandomly();
+            cluster.stabilise();
+
+            cluster.addNodesAndStabilise(1);
+            final Cluster.ClusterNode newNode = cluster.clusterNodes.get(cluster.clusterNodes.size() - 1);
+            final PublishClusterStateStats newNodePublishStats = newNode.coordinator.stats().getPublishStats();
+            // initial cluster state send when joining
+            assertEquals(1L, newNodePublishStats.getFullClusterStateReceivedCount());
+            // no reconfiguration
+            assertEquals(0, newNodePublishStats.getCompatibleClusterStateDiffReceivedCount());
+            assertEquals(0L, newNodePublishStats.getIncompatibleClusterStateDiffReceivedCount());
+        }
+    }
+
+    @Override
+    protected CoordinatorStrategy getCoordinatorStrategy() {
+        return new AtomicRegisterCoordinatorStrategy(keyedAtomicRegister);
+    }
+
+    record HeartBeat(DiscoveryNode leader, long absoluteTimeInMillis) {
+        long timeSinceLastHeartbeatInMillis(long nowInMillis) {
+            return nowInMillis - absoluteTimeInMillis;
+        }
+    }
+
+    static class AtomicRegisterHeartbeatService implements LeaderHeartbeatService {
```

Comment:
naming nit, sometimes it's `Heartbeat` and sometimes `HeartBeat`. It's fine as a single word (like `Metadata`) so I prefer that, although I would prefer something involving `Liveness` even more :)